src/errors/exception.ts

```diff
@@ -0,0 +1,59 @@
+import { utils } from 'ethers';
+import { SWIVEL_ABI } from '../constants/index.js';
+import { Exception, ExceptionResult, isStaticCallError, isUnpredictableGasLimitError } from './types.js';
+
+// swivel abi for decoding exception data
+const SWIVEL_INTERFACE = new utils.Interface(SWIVEL_ABI);
+
+// abi fragment name of the custom error
+const SWIVEL_ERROR_FRAGMENT = 'Exception';
+
+/**
+ * Parses the custom error data from an `UNPREDICTABLE_GAS_LIMIT` or `CALL_EXCEPTION` error.
+ *
+ * @param error - an ethers error/rejection reason that may contain custom error data
+ * @returns an {@link Exception} if the error contains custom error data, `undefined` otherwise
+ *
+ * @example
+ */
+export const parseException = (error: unknown): Exception | undefined => {
+
+    // parse custom error data from an unpredictable gas limit error
+    if (isUnpredictableGasLimitError(error)) {
+
+        try {
+
+            const [code, amount, amountExpected, address, addressExpected] = SWIVEL_INTERFACE.decodeErrorResult(
+                SWIVEL_ERROR_FRAGMENT,
+                error.error.error.error.data,
```

Comment:
yo dawg..

Reply:
also could the depth of refinements there ever change -- i'm just not sure what's placing 3 moar errors on error before you get to data